crates/compiler/unify/src/unify.rs

```diff
@@ -1451,19 +1442,60 @@ fn separate_union_lambdas<M: MetaCollector>(
                             //
                             // If they are not unifiable, that means the two lambdas must be
                             // different (since they have different capture sets), and so we don't
-                            // want to merge the variables.
-                            let variables_are_unifiable = env.with_outcome_only(|env| {
-                                unify_pool::<NoCollector>(env, pool, var1, var2, mode)
-                                    .mismatches
-                                    .is_empty()
-                            });
-
-                            if !variables_are_unifiable {
-                                continue 'try_next_right;
-                            }
```

Comment:
Of course, if this had worked right, it would have been a terrible idea - this is quadratic in its operation at best, but without the bug this PR fixes, it's exponential. The lesser of two evils is to suffer a clone of subs and rollback if we need to.

maybe in the future we can do something fancier - for example, one heuristic that I think is promising is that two lambdas of the same name are very likely to unify and not need a niche. So, we can snapshot once when unifying at the start of two lambdas, and if there is any error, then rollback the whole thing.